src/metabase/models/user_parameter_value.clj

```diff
@@ -30,31 +33,49 @@
   {:value {:in  mi/json-in
            :out json-out}})
 
-(mu/defn batched-upsert!
+(defn batched-upsert!
   "Delete param with nil value and upsert the rest."
+  [parameters]
+  (let [{to-delete true
+         to-upsert false} (group-by #(and (nil? (:value %)) (nil? (:default %))) parameters)]
+    (t2/with-transaction [_conn]
+      (doseq [item to-upsert]
+        (or (pos? (t2/update! :model/UserParameterValue (select-keys item [:user_id :dashboard_id :parameter_id])
+                              {:value (:value item)}))
+            (t2/insert! :model/UserParameterValue (select-keys item [:user_id :dashboard_id :parameter_id :value]))))
+      (when (seq to-delete)
+        (t2/delete! :model/UserParameterValue {:where (into [:or] (for [p to-delete]
+                                                                    [:and
+                                                                     [:= :user_id (:user_id p)]
+                                                                     [:= :dashboard_id (:dashboard_id p)]
+                                                                     [:= :parameter_id (:parameter_id p)]]))})))))
+
+(defonce ^:private user-parameter-value-queue
+  (delay (grouper/start!
+          (fn [inputs]
+            (try
+              (batched-upsert!
+               (->> (for [input     inputs
+                          parameter (:parameters input)]
+                      {:user_id      (:user-id input)
+                       :dashboard_id (:dashboard-id input)
+                       :parameter_id (:id parameter)
+                       :value        (:value parameter)
+                       :default      (:default parameter)})
+                    (m/distinct-by (juxt :user_id :dashboard_id :parameter_id))))
+              (catch Exception e
+                (log/error e "Error saving user parameters for a dashboard"))))
+          :capacity 50
+          :interval 100)))
```

Comment:
100 milliseconds seems like a short interval to me. Do we need this to dump to the db 10 times a second? I would think a longer window was fine for our purposes.